MyThreadPool/MyThreadPool/MyThreadPool.cs

```diff
@@ -0,0 +1,311 @@
+            }
+            _taskQueue.Enqueue(task);
+            Monitor.Pulse(_lock);
+        }
+    }
+
+    /// <summary>
+    /// The worker method for the threads in the pool.
+    /// Continuously executes tasks from the queue until the pool is shut down and the queue is empty.
+    /// </summary>
+    private void Work()
+    {
+        while (true)
+        {
+            Action task = null;
+            lock (_lock)
+            {
+                while (_taskQueue.Count == 0 && !_isShutdown)
+                {
+                    Monitor.Wait(_lock);
+                }
+                if (_isShutdown && _taskQueue.Count == 0)
+                {
+                    return;
+                }
+                if (_taskQueue.Count > 0)
+                {
+                    task = _taskQueue.Dequeue();
+                }
+            }
+            task?.Invoke();
+        }
+    }
+
+    /// <summary>
+    /// Submits a task to be executed by the thread pool.
+    /// The task is represented as a function returning a result of type <typeparamref name="TResult"/>.
+    /// </summary>
+    /// <typeparam name="TResult">The type of result produced by the task.</typeparam>
+    /// <param name="task">The function representing the task to execute.</param>
+    /// <returns>An <see cref="IMyTask{TResult}"/> representing the submitted task.</returns>
+    /// <exception cref="ArgumentNullException">Thrown if the provided task is null.</exception>
+    /// <exception cref="InvalidOperationException">Thrown if the thread pool has been shut down.</exception>
+    public IMyTask<TResult> SubmitTask<TResult>(Func<TResult> task)
+    {
+        if (task == null)
+        {
+            throw new ArgumentNullException(nameof(task));
+        }
+        lock (_lock)
+        {
+            if (_isShutdown)
+            {
+                throw new InvalidOperationException("ThreadPool has been shut down. Cannot accept new tasks.");
+            }
+            var myTask = new MyTask<TResult>(task, this, _cancellationTokenSource.Token); 
+            _taskQueue.Enqueue(() => myTask.Execute());
+            Monitor.Pulse(_lock);
+            return myTask;
+        }
+    }
+
+    /// <summary>
+    /// Shuts down the thread pool and waits for all threads to complete.
+    /// No new tasks are accepted once shutdown is initiated.
+    /// </summary>
+    public void Shutdown()
+    {
+        lock (_lock)
+        {
+            _isShutdown = true;
+            _cancellationTokenSource.Cancel();
+            Monitor.PulseAll(_lock);
+        }
+
+        foreach (var thread in _threads)
+        {
+            thread.Join();
+        }
+    }
+
+    /// <summary>
+    /// Represents a task that can be executed by the thread pool.
+    /// </summary>
+    /// <typeparam name="TResult">The type of the result produced by the task.</typeparam>
+    public class MyTask<TResult> : IMyTask<TResult>
```

Comment:
```suggestion
    private class MyTask<TResult> : IMyTask<TResult>
```
зачем ему быть public, для этого есть IMyTask